crates/vm/src/system/memory/controller/mod.rs

```diff
@@ -463,46 +440,151 @@ impl<F: PrimeField32> MemoryController<F> {
     {
-        if self.final_state.is_some() {
-            return;
-        }
-        todo!();
-
+        let touched_blocks = self.memory.touched_blocks().collect::<Vec<_>>();
+        // First, let's split everything into blocks of ALIGN
+        let aligned_final_memory = {
+            let mut final_memory = TimestampedEquipartition::<F, RV32_REGISTER_NUM_LIMBS>::new();
+            for ((addr_space, ptr), metadata) in touched_blocks {
+                if metadata.block_size == AccessMetadata::OCCUPIED {
+                    continue;
+                }
+                assert!(
+                    metadata.block_size >= RV32_REGISTER_NUM_LIMBS as u32,
+                    "How come some block size is less than RV32_REGISTER_NUM_LIMBS, isn't this the alignment?"
+                );
+                let values = (0..metadata.block_size)
+                    .map(|i| self.memory.data.get_f::<F>(addr_space, ptr + i))
+                    .collect::<Vec<_>>();
+                for start in (0..metadata.block_size).step_by(RV32_REGISTER_NUM_LIMBS) {
+                    final_memory.insert(
+                        (addr_space, ptr + start),
+                        TimestampedValues {
+                            timestamp: metadata.timestamp,
+                            values: values
+                                [(start as usize)..(start as usize + RV32_REGISTER_NUM_LIMBS)]
+                                .try_into()
+                                .unwrap(),
+                        },
+                    );
+                }
+                self.memory.execute_splits::<RV32_REGISTER_NUM_LIMBS, true>(
+                    MemoryAddress::new(addr_space, ptr),
+                    &values,
+                    metadata.timestamp,
+                );
+            }
+            final_memory
+        };
         match &mut self.interface_chip {
             MemoryInterface::Volatile { boundary_chip } => {
-                // let final_memory = offline_memory.finalize::<1>(&mut self.access_adapters);
-                // boundary_chip.finalize(final_memory);
-                // self.final_state = Some(FinalState::Volatile(VolatileFinalState::default()));
+                let mut final_memory = TimestampedEquipartition::<F, 1>::new();
+                for ((addr_space, ptr), ts_values) in aligned_final_memory {
+                    let TimestampedValues {
+                        timestamp,
+                        mut values,
+                    } = ts_values;
+                    for (i, &v) in values.iter().enumerate() {
+                        final_memory.insert(
+                            (addr_space, ptr + i as u32),
+                            TimestampedValues {
+                                timestamp,
+                                values: [v],
+                            },
+                        );
+                    }
+                    self.memory.execute_splits::<1, false>(
```

Comment:
does the size in `execute_splits` need to be const generic? life may be easier for you if it isn't in the future